# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/jit/Bailouts.h
# Commit: 7f75f8b705e7
# Full Hash: 7f75f8b705e72090c3faa76aa8bac440c6d0fde6
# Author: Iain Ireland <iireland@mozilla.com>
# Date: 2022-03-31 09:35:41
# Regressor Bug: 885514
# File Overlap Count: 2
# Description:
#   Bug 885514: Support bailouts for finally r=jandem
#   
#   This is mostly modeled after the catch code. The main difference is that finally expects to resume in code where there are two additional values on the stack, so we store the exception in the ExceptionBailoutInfo and write it into the stack frame in BaselineStackBuilder.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D142118
# ==============================================================================

diff -r 8cd3f1a267ae -r 7f75f8b705e7 js/src/jit/Bailouts.h
--- a/js/src/jit/Bailouts.h	Wed Mar 30 23:19:47 2022 +0000
+++ b/js/src/jit/Bailouts.h	Wed Mar 30 23:19:47 2022 +0000
@@ -192,13 +192,22 @@
   size_t frameNo_;
   jsbytecode* resumePC_;
   size_t numExprSlots_;
+  bool isFinally_ = false;
+  RootedValue finallyException_;
 
  public:
-  ExceptionBailoutInfo(size_t frameNo, jsbytecode* resumePC,
+  ExceptionBailoutInfo(JSContext* cx, size_t frameNo, jsbytecode* resumePC,
                        size_t numExprSlots)
-      : frameNo_(frameNo), resumePC_(resumePC), numExprSlots_(numExprSlots) {}
+      : frameNo_(frameNo),
+        resumePC_(resumePC),
+        numExprSlots_(numExprSlots),
+        finallyException_(cx) {}
 
-  ExceptionBailoutInfo() : frameNo_(0), resumePC_(nullptr), numExprSlots_(0) {}
+  explicit ExceptionBailoutInfo(JSContext* cx)
+      : frameNo_(0),
+        resumePC_(nullptr),
+        numExprSlots_(0),
+        finallyException_(cx) {}
 
   bool catchingException() const { return !!resumePC_; }
   bool propagatingIonExceptionForDebugMode() const { return !resumePC_; }
@@ -215,6 +224,17 @@
     MOZ_ASSERT(catchingException());
     return numExprSlots_;
   }
+
+  bool isFinally() const { return isFinally_; }
+  void setFinallyException(JS::Value& exception) {
+    MOZ_ASSERT(!isFinally());
+    isFinally_ = true;
+    finallyException_ = exception;
+  }
+  HandleValue finallyException() const {
+    MOZ_ASSERT(isFinally());
+    return finallyException_;
+  }
 };
 
 // Called from the exception handler to enter a catch or finally block.